api/common/api-profile.go

```diff
@@ -0,0 +1,441 @@
+
+// AttributesToApi converts a KC profile attribute to its API version
+func AttributesToApi(attrbs []kc.ProfileAttrbRepresentation, frontend string, requesterType string) []ProfileAttributeRepresentation {
+	if attrbs == nil {
+		return nil
+	}
+	if len(attrbs) == 0 {
+		return make([]ProfileAttributeRepresentation, 0)
+	}
+	var res []ProfileAttributeRepresentation
+	for _, attrb := range attrbs {
+		var newValue = AttributeToApi(attrb, frontend, requesterType)
+		if newValue != nil {
+			res = append(res, *newValue)
+		}
+	}
+	return res
+}
+
+func AttributeToApi(attrb kc.ProfileAttrbRepresentation, frontend string, requesterType string) *ProfileAttributeRepresentation {
+	if attrb.Permissions != nil && !validation.IsStringInSlice(attrb.Permissions.Edit, requesterType) {
+		// User has no permission to edit this field
+		// Register should not be aware of its existence
+		return nil
+	}
+	// By default, attributes are not shown for a given frontend
+	if frontend != "" && !attrb.IsAnnotationTrue(frontend) {
+		// Attribute is not declared to be used by the given frontend
+		return nil
+	}
+	var required = isAttributeRequired(attrb, requesterType)
+	return &ProfileAttributeRepresentation{
+		Name:        attrb.Name,
+		DisplayName: attrb.DisplayName,
+		Group:       attrb.Group,
+		Required:    &required,
+		Validations: ValidationsToApi(attrb.Validations),
+	}
+}
+
+func isAttributeRequired(attrb kc.ProfileAttrbRepresentation, requesterType string) bool {
+	if attrb.Required == nil {
+		return false
+	}
+	return validation.IsStringInSlice(attrb.Required.Roles, requesterType)
+}
+
+// ValidationsToApi converts KC validators
+func ValidationsToApi(validations kc.ProfileAttrbValidationRepresentation) map[string]ProfileAttrbValidatorRepresentation {
+	if validations == nil {
+		return nil
+	}
+	var res = make(map[string]ProfileAttrbValidatorRepresentation)
+	for k, v := range validations {
+		res[k] = ToValidator(v)
+	}
+
+	return res
+}
+
+// ToValidator converts a KC validator
+func ToValidator(validator kc.ProfileAttrValidatorRepresentation) ProfileAttrbValidatorRepresentation {
+	if validator == nil {
+		return nil
+	}
+	var res = make(ProfileAttrbValidatorRepresentation)
+	for k, v := range validator {
+		res[k] = v
+	}
+	return res
+}
+
+// GroupsToApi converts KC users profile groups to its API version
+func GroupsToApi(groups []kc.ProfileGroupRepresentation) []ProfileGroupRepresentation {
+	if groups == nil {
+		return nil
+	}
+	if len(groups) == 0 {
+		return make([]ProfileGroupRepresentation, 0)
+	}
+	var res []ProfileGroupRepresentation
+	for _, group := range groups {
+		res = append(res, ProfileGroupRepresentation{
+			Name:               group.Name,
+			DisplayHeader:      group.DisplayHeader,
+			DisplayDescription: group.DisplayDescription,
+			Annotations:        group.Annotations,
+		})
+	}
+	return res
+}
+
+// DefaultProfile returns a default profile
+func DefaultProfile(realmName string) (kc.UserProfileRepresentation, error) {
```

Comment:
I would I preffered to have this in keycloak-client... but I consider keycloak-client as a golang client for Keycloak. If I move the default profile in keycloak-client, it would be a cloudtrust-specific default configuration, that's why I kept it here